File: app/src/main/java/com/veganbeauty/app/features/shop/home/ShopHomeViewModel.java
package com.veganbeauty.app.features.shop.home;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.veganbeauty.app.R;
import com.veganbeauty.app.core.base.RootieViewModel;
import com.veganbeauty.app.data.local.entities.ProductEntity;
import com.veganbeauty.app.data.repository.ProductRepository;
import com.veganbeauty.app.features.shop.home.models.BannerUiModel;
import com.veganbeauty.app.features.shop.home.models.CategoryUiModel;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class ShopHomeViewModel extends RootieViewModel {

    private final ProductRepository productRepository;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();

    private final MutableLiveData<List<BannerUiModel>> _banners = new MutableLiveData<>(new ArrayList<>());
    public final LiveData<List<BannerUiModel>> banners = _banners;

    private final MutableLiveData<List<CategoryUiModel>> _categories = new MutableLiveData<>(new ArrayList<>());
    public final LiveData<List<CategoryUiModel>> categories = _categories;

    private final MutableLiveData<List<ProductEntity>> _suggestedProducts = new MutableLiveData<>(new ArrayList<>());
    public final LiveData<List<ProductEntity>> suggestedProducts = _suggestedProducts;

    private androidx.lifecycle.Observer<List<ProductEntity>> observer;
    private LiveData<List<ProductEntity>> allProductsLive;

    public ShopHomeViewModel(ProductRepository productRepository) {
        this.productRepository = productRepository;
        applyDefaultUi();
        refreshProducts();
        observeProductCounts();
    }

    public void preloadUiData(List<ProductEntity> localProducts) {
        List<ProductEntity> products = localProducts != null ? localProducts : new ArrayList<>();
        _banners.postValue(buildBanners());
        _categories.postValue(buildCategories(products));
        _suggestedProducts.postValue(products);
    }

    private void applyDefaultUi() {
        _banners.setValue(buildBanners());
        _categories.setValue(buildCategories(new ArrayList<>()));
    }

    private static List<BannerUiModel> buildBanners() {
        List<BannerUiModel> bannerList = new ArrayList<>();
        bannerList.add(new BannerUiModel("1", R.drawable.shop_banner_1));
        bannerList.add(new BannerUiModel("2", R.drawable.shop_banner_2));
        bannerList.add(new BannerUiModel("3", R.drawable.shop_banner_3));
        return bannerList;
    }

    private static List<CategoryUiModel> buildCategories(List<ProductEntity> products) {
        return ShopCategoryHelper.buildHomeCategories(products);
    }

    private void observeProductCounts() {
        allProductsLive = androidx.lifecycle.FlowLiveDataConversions.asLiveData(productRepository.getAllProducts());

        observer = products -> {
            if (products == null || products.isEmpty()) return;
            executor.execute(() -> {
                try {
                    _suggestedProducts.postValue(products);
                    _categories.postValue(buildCategories(products));
                } catch (Exception e) {
                    e.printStackTrace();
                }
            });
        };
        allProductsLive.observeForever(observer);
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        if (allProductsLive != null && observer != null) {
            allProductsLive.removeObserver(observer);
        }
        executor.shutdown();
    }

    private void refreshProducts() {
        executor.execute(() -> {
            try {
                List<ProductEntity> localProducts = productRepository.getLocalJsonReader().getAllProducts();
                if (localProducts != null && !localProducts.isEmpty()) {
                    List<ProductEntity> current = _suggestedProducts.getValue();
                    if (current == null || current.isEmpty()) {
                        preloadUiData(localProducts);
                    }
                    productRepository.syncAllProductImagesFromAssets(localProducts);
                }
                productRepository.refreshProducts(true);
            } catch (Exception e) {
                e.printStackTrace();
            }
        });
    }
}

File: app/src/main/java/com/veganbeauty/app/features/shop/home/ShopHomeViewModel.kt
package com.veganbeauty.app.features.shop.home

import androidx.lifecycle.asLiveData
import androidx.lifecycle.viewModelScope
import com.veganbeauty.app.core.base.RootieViewModel
import com.veganbeauty.app.data.repository.ProductRepository
import com.veganbeauty.app.features.shop.home.models.BannerUiModel
import com.veganbeauty.app.features.shop.home.models.CategoryUiModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class ShopHomeViewModel(private val productRepository: ProductRepository) : RootieViewModel() {

    // Danh sách banner (Mock data tạm thời cho UI)
    private val _banners = MutableStateFlow<List<BannerUiModel>>(emptyList())
    val banners: StateFlow<List<BannerUiModel>> = _banners.asStateFlow()

    // Danh sách category (Mock data tạm thời cho UI)
    private val _categories = MutableStateFlow<List<CategoryUiModel>>(emptyList())
    val categories: StateFlow<List<CategoryUiModel>> = _categories.asStateFlow()

    // Danh sách sản phẩm gợi ý lấy từ Repository (sử dụng ProductEntity)
    val suggestedProducts = productRepository.allProducts.asLiveData()

    init {
        loadMockData()
        refreshProducts()
        observeProductCounts()
    }

    private fun loadMockData() {
        // Mock dữ liệu Banner (Sử dụng 3 ảnh organic đã gen)
        _banners.value = listOf(
            BannerUiModel("1", com.veganbeauty.app.R.drawable.shop_banner_1), 
            BannerUiModel("2", com.veganbeauty.app.R.drawable.shop_banner_2),
            BannerUiModel("3", com.veganbeauty.app.R.drawable.shop_banner_3)
        )

        // Mock dữ liệu Danh mục ban đầu trước khi load sản phẩm
        _categories.value = listOf(
            CategoryUiModel("c1", "Chăm sóc da", com.veganbeauty.app.R.drawable.ic_skincare, 0),
            CategoryUiModel("c2", "Tắm & Dưỡng thể", com.veganbeauty.app.R.drawable.ic_shower, 0),
            CategoryUiModel("c4", "Dưỡng môi", com.veganbeauty.app.R.drawable.ic_lips, 0),
            CategoryUiModel("c5", "Combo/Giftbox", com.veganbeauty.app.R.drawable.ic_combo, 0)
        )
    }

    private fun observeProductCounts() {
        viewModelScope.launch {
            productRepository.allProducts.collect { products ->
                val skincareIds = listOf(
                    "f5877af6a55f88bcf57c17b4", "389971929086b2ce7fba9dd0",
                    "36cbf3f5c4b7a299ce2a2d0c", "4e20d6bbc1203015ee2ecd48",
                    "b1b6cd208332d4f1e015a26c", "7667d982515426a9d88b787b",
                    "bb88a3306cf95af20d073594", "9882d5fa14c74dd053e17f33",
                    "c211afa24702f5d1ff86fe42"
                )
                val bodyIds = listOf(
                    "7c70e845e829b374e57ee7b1", "b703bb813e660aa88076ee5a",
                    "8fce5340c618672aa1ae7fb3"
                )
                val lipsIds = listOf("755731e01d8c579c633ae4d2", "ded17e0716783c133b1a5b9a")
                val comboIds = listOf(
                    "7176b5e7966be88daf95cfd4", "f40c1f05dcf4059f25fb89a1",
                    "e0754dabb88699e92481e123", "bd1c0ff76b19b1b5a3130a79"
                )

                var skincareCount = 0
                var bodyCount = 0
                var lipsCount = 0
                var comboCount = 0

                products.forEach { product ->
                    val productCategoryIds = product.categoryIds.split(",")
                    if (productCategoryIds.any { it in skincareIds }) skincareCount++
                    if (productCategoryIds.any { it in bodyIds }) bodyCount++
                    if (productCategoryIds.any { it in lipsIds }) lipsCount++
                    if (productCategoryIds.any { it in comboIds }) comboCount++
                }

                _categories.value = listOf(
                    CategoryUiModel("c1", "Chăm sóc da", com.veganbeauty.app.R.drawable.ic_skincare, skincareCount),
                    CategoryUiModel("c2", "Tắm & Dưỡng thể", com.veganbeauty.app.R.drawable.ic_shower, bodyCount),
                    CategoryUiModel("c4", "Dưỡng môi", com.veganbeauty.app.R.drawable.ic_lips, lipsCount),
                    CategoryUiModel("c5", "Combo/Giftbox", com.veganbeauty.app.R.drawable.ic_combo, comboCount)
                )
            }
        }
    }

    private fun refreshProducts() {
        viewModelScope.launch {
            try {
                productRepository.refreshProducts()
            } catch (e: Exception) {
                // Xử lý lỗi nếu cần
            }
        }
    }
}
